Fix OOM detection for pods with restartPolicy: Never

tkukushkin opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Current approach of detection OOMs is based on `RestartCount` and `status.containerStatuses.lastState` of pod. But if pod doesn't restart (`restartPolicy: Never`), pod doesn't have `lastState` and `RestartCount` isn't increased.

I changed it to check `status.containerStatuses.state.terminated` instead. If container's status changes from running to terminated with `OOMKilled` reason, we record OOM. I believe this approach works for regular pods that can restart and works for not restarting pods as well.

If it is regular pod which restarts:
1. Container is running (`status.containerStatuses.state.terminated == nil`)
2. Container is OOM killed (`status.containerStatuses.state.terminated.reason == "OOMKilled"`)
3. We record OOM
4. Container restarts, it's `state.terminated` resets (`status.containerStatuses.state.terminated == nil`)
6. See 1.

If it is pod which doesn't restart:
1. Container is running (`status.containerStatuses.state.terminated == nil`)
2. Container is OOM killed (`status.containerStatuses.state.terminated.reason == "OOMKilled"`)
3. We record OOM

#### Which issue(s) this PR fixes:

Fixes #6236

#### Does this PR introduce a user-facing change?

```release-note
- Fixed OOM detection for pods with `restartPolicy: Never`
```

adrianmoisey commented:
/ok-to-test

voelzmo commented:
I'm not sure if it is safe to switch entirely to `.status.state` and disregard the `restartCount` event for restartable Containers.

Looking at the docs for https://pkg.go.dev/k8s.io/client-go@v0.33.0/tools/cache#ResourceEventHandler it states that events may be combined if they happen quickly after another
> `OnUpdate` is called when an object is modified. Note that oldObj is the last known state of the object-- it is possible that several changes were combined together, so you can't use this to see every single change. OnUpdate is also called when a re-list happens, and it will get called even if nothing changed. This is useful for periodically evaluating or syncing something.

After merging the PR, the oom observer only considers Pod update events that updated `.status.state.terminated` from `nil` in the `oldObject` to non-nil in the `object`. This works iff:
* in between two OOMKills an update event from `.status.state.terminated` to `.status.state.running` is observed – so the Pod has to be running "long enough", and I don't know how long that might be
* for Containers which _are restartable_, the Controller doesn't restart "too quickly", such that the observer can get the event where `.status.state.terminated` is non-nil

Maybe we need to consider the Pod's [restartPolicy](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy) and choose which fields to read in the Container's status?

What do the others think, @raywainman @adrianmoisey @omerap12?

tkukushkin commented:
> events may be combined if they happen quickly after another

If events may be combined, this approach may not work in some edge cases, I didn't know about this behaviour. Isn't it possible to disable it somehow?

By the way, we've been using this approach in production for several months and haven't noticed any problems like that, but of course we can't be sure it **always** works like expected.

Also, I could be wrong, but it seems to me that the `restartCount` tracking has another flaw. `restartCount` is increased only when container is actually restarted, but there is backoff time as weel, so OOM may be recorded later, sometimes much later.

voelzmo commented:
> OOM may be recorded later, sometimes much later.

That definitely is an issue! I think we had people report instances where OOMKill events had been dropped with 
`OOM event will be discarded - it is too old` which might happen if the aggregation window changed between OOMKill and the Pod restart. Not sure if there might be other causes for this to occur, though.

omerap12 commented:
> Maybe we need to consider the Pod's [restartPolicy](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy) and choose which fields to read in the Container's status?

Yeah I agree, I think we should define the behavior based on the `restartPolicy`:

For `restartPolicy: Never`:
- Use `containerStatus.State.Terminated` check

For `restartPolicy: Always/OnFailure`:
- Primary: Use `RestartCount` +` LastTerminationState` check
- Fallback: Use `State.Terminated` check
  (to catch OOMs that might be missed due to backoff/aggregation window issues)

tkukushkin committed:
Fix OOM detection for jobs with restartPolicy Never

adrianmoisey commented on vertical-pod-autoscaler/pkg/recommender/input/oom/observer.go:
```diff
@@ -148,33 +148,63 @@ func (o *observer) OnUpdate(oldObj, newObj any) {
+		isNewOOM := containerStatus.State.Terminated != nil &&
+			containerStatus.State.Terminated.Reason == "OOMKilled" &&
+			(oldStatus.State.Terminated == nil ||
+				containerStatus.RestartCount > oldStatus.RestartCount)
```
Since both `oldStatus.State.Terminated == nil` and `containerStatus.RestartCount > oldStatus.RestartCount` are used in isNewOOM and isPreviousOOM, could this be moved out to their own variables?

It may make this easier to read

tkukushkin replied:
Done in 5e930a4 — extracted `oldNotTerminated` and `restartCountIncreased` as shared variables.

adrianmoisey commented on vertical-pod-autoscaler/pkg/recommender/input/oom/observer.go:
```diff
@@ -148,33 +148,63 @@ func (o *observer) OnUpdate(oldObj, newObj any) {
+		oldSpec := findSpec(containerStatus.Name, oldPod.Spec.Containers)
+		if oldSpec == nil {
+			continue
 		}
```
What is the purpose of this check?

adrianmoisey replied:
Oh, is this testing that previous pod has the container of the current pod?
If so, can you rename the variable and add a comment?

tkukushkin replied:
Good catch. The check is actually dead code: since 204ac568 ("Also read resources from containerStatus in oom observer") the memory is read via `resourcehelpers.ContainerRequestsAndLimits(name, pod)` instead of `oldSpec.Resources`, and the `oldSpec` variable hasn't been used for anything other than the nil check itself. Since we also already `continue` when `oldStatus == nil` (and a container present in `Status.ContainerStatuses` has to exist in `Spec.Containers`), the check is redundant.

Removed it (and the now-unused `findSpec` helper) in 5e930a4.

While touching this block I also noticed `ContainerRequestsAndLimits` was always being called with `oldPod`, which isn't quite right now that we have two branches: for `isNewOOM` the OOM just happened on the container currently in `newPod` (its resources should be read from there), and for `isPreviousOOM` the OOM happened on the pre-restart instance, whose resources are those of `oldPod`. Fixed in the same commit and adjusted the tests.

tkukushkin committed:
Address review comments: extract shared conditions, pick pod for OOM resources

- Extract oldNotTerminated and restartCountIncreased into variables to
  deduplicate the isNewOOM/isPreviousOOM conditions.
- Read requests from newPod for isNewOOM (OOM happened in the current
  terminated container) and from oldPod for isPreviousOOM (OOM happened
  in the previous container before the observed restart).
- Drop the oldSpec nil-check: the variable has been unused since the
  switch to ContainerRequestsAndLimits, and oldStatus already proves the
  container exists.
- Fix apiv1 -> corev1 leftovers from the importas rename.
- Fix 'direcly' typo.

tkukushkin committed:
Add tests for isPreviousOOM, fast-restart, and non-OOM terminations

- TestOOMReceivedPreviousOOM: covers the Running->Running restart branch
  with LastTerminationState OOMKilled, plus a case verifying resources
  are read from oldPod.
- TestOOMReceivedFastRestart: covers the fast-restart branch of
  isNewOOM (Terminated->Terminated with increased RestartCount).
- TestNonOOMTerminationIgnored: verifies that Terminated with non-OOM
  reason and Running->Running with non-OOM LastTerminationState are
  both ignored.

adrianmoisey commented:
/retest

adrianmoisey commented:
From my understanding and testing, this seems to improve OOM detection correctly.

/cc maxcao13 

Max, any chance you have some cycles to look a this too?

maxcao13 commented on vertical-pod-autoscaler/pkg/recommender/input/oom/observer.go:
```diff
@@ -148,33 +139,69 @@ func (o *observer) OnUpdate(oldObj, newObj any) {
+		// Terminated state and change directly from Running
+		// to Running with increased RestartCount. In this
+		// case we check LastTerminationState.
+		isPreviousOOM := containerStatus.State.Running != nil &&
+			oldNotTerminated &&
+			restartCountIncreased &&
 			containerStatus.LastTerminationState.Terminated != nil &&
```
From manually testing and debug statements, I think there's a bug if the container goes into CrashLoopBackoff (waiting state).

1. Container A transitions (observed) `Running` -> `Terminated (OOMKill)`, restartCount 0 -> 0, so `isNewOOM` will trigger and record an OOM sample.
2. Container A transitions (observed) `Terminated` -> `Waiting (CrashLoopBackoff)`, restartCount 0 -> 0, and both the previousOom and newOom will skip since state is now `Waiting` and neither `Terminated` or `Running`.
3. Container A transitions (observed) `Waiting` -> `Running` , restartCount 0 -> 1, but here the `lastState.terminated.reason==OOMKill` so `isPreviousOOM` triggers, because `isPreviousOOM` checks if the previous containerStatus was NOT `Terminated`, and here the previous status was `Waiting`, which records another OOM sample.

So I believe we are double counting from 1 oomkill. Here in the example, we are actually observing every state that actually happens. I think we could fix this if we change the check in `isPreviousOOM` to check if `lastState == Running` and not just `lastState != Terminated` since there exists the `Waiting` container state.

tkukushkin replied:
Confirmed. Reproduced in a test cluster — the container does go Waiting(CrashLoopBackOff) -> Running while keeping `lastState.terminated.reason=OOMKilled`, so `isPreviousOOM` fires a second time for an OOM already recorded by `isNewOOM` on the preceding Running -> Terminated transition.

Applied the suggested fix in ad5fb271 — `isPreviousOOM` now requires `oldStatus.State.Running != nil` instead of just `oldStatus.State.Terminated == nil`. Added `TestOOMNotDoubleCountedAcrossCrashLoopBackoff` that covers exactly this sequence.

maxcao13 commented on vertical-pod-autoscaler/pkg/recommender/input/oom/observer_test.go:
```diff
@@ -181,18 +179,291 @@ func TestOOMReceived(t *testing.T) {
+	}
+}
 
+func TestOOMStateAfterTerminatedState(t *testing.T) {
```
Can we add one more test which is just to check that a simulated pod informer re-list won't double count an OOMkill?

i.e. if a container 
1. went running -> OOMKill -> running such that `lastState.terminatedReason` is now `OOMKilled` will trigger an OOMKill observer event in the oomchannel
2. Then we simulate an informer relist, which causes an update to list the same pod, but lastState.terminatedReason is still `OOMKilled`. Can we test that this wouldn't trigger a new `OOMKill` event to be sent to the same oomobserver channel?

tkukushkin replied:
Added as `TestOOMNotDoubleCountedOnInformerRelist` in ad5fb271 — feeds the observer identical old/new pod state and asserts the channel stays empty.

tkukushkin committed:
Fix isPreviousOOM double-counting across CrashLoopBackOff

A container that OOM-kills and enters CrashLoopBackOff transitions
Waiting -> Running on the next restart, while keeping
lastState.terminated set to the previous OOMKill. The old isPreviousOOM
check only required oldStatus.State.Terminated == nil, so it fired
on that transition and double-counted an OOM that had already been
recorded as isNewOOM on the preceding Running -> Terminated update.

Require oldStatus.State.Running != nil instead. Add tests for the
CrashLoopBackOff transition and for an informer relist emitting the
same state twice.

maxcao13 reviewed (approved):
/lgtm

thanks for the work on this :-)

omerap12 reviewed (commented):
Thanks for this! Overall, this LGTM, but there is quite a bit of code repetition in the test file. Can we merge these test functions into a single event-driven test function?

tkukushkin committed:
Merge observer OnUpdate tests into one table-driven test

Collapse the seven separate Test* functions covering observer.OnUpdate
into a single TestOOMObserverOnUpdate table, add a mustNewPod helper,
and factor the common pod metadata/spec into a shared podYamlHeader
constant so each fixture declares only the containerStatus fields that
distinguish its scenario.

tkukushkin commented:
@omerap12 done in cd7bb887 — merged the seven `OnUpdate` tests into a single table-driven `TestOOMObserverOnUpdate` and added a `mustNewPod` helper plus a shared `podYamlHeader` constant so each fixture only declares the containerStatus fields that distinguish its scenario. Test file shrank from 705 to 430 lines.

omerap12 commented on vertical-pod-autoscaler/pkg/recommender/input/oom/observer_test.go:
```diff
@@ -491,79 +145,161 @@ status:
+	if err != nil {
+		t.Fatalf("failed to parse pod YAML: %v", err)
+	}
+	return pod
```
Since `newPod` function being used only in `mustNewPod` we can merged those as well:

```golang
func mustNewPod(t *testing.T, yaml string) *corev1.Pod {
	t.Helper()
	decode := codecs.UniversalDeserializer().Decode
	obj, _, err := decode([]byte(yaml), nil, nil)
	if err != nil {
		t.Fatalf("failed to parse pod YAML: %v", err)
	}
	return obj.(*corev1.Pod)
}
```

and then delete `newPod` entirely

tkukushkin replied:
Done in 6910f909 — inlined the decode logic into `mustNewPod` and removed `newPod`.

tkukushkin committed:
Inline newPod into mustNewPod

newPod has no other callers now that all observer OnUpdate tests use
mustNewPod, so merge the decode logic directly into the helper.

omerap12 reviewed (approved):
Thanks! 
/lgtm
/approve
/label tide/merge-method-squash

tkukushkin commented:
/retest
